### ISSUE TITLE
Getting an internal server error on my deployed app

### ISSUE THREAD

**Issue by TheOnlyArtz** [novatorem/novatorem#7]
Hey, I've followed every step from the SetUp.md (for the 3rd time now) and I'm getting 500 on my app:
novatorem.theonlyartz.vercel.app/api/spotify

build logs (although it looks just fine):
```
16:35:01.421  	Cloning github.com/TheOnlyArtz/novatorem (Branch: master, Commit: 7eebe7e)
16:35:01.982  	Cloning completed in 561ms
16:35:01.983  	Analyzing source code...
16:35:02.699  	Uploading build outputs...
16:35:04.176  	Installing build runtime...
16:35:05.761  	Build runtime installed: 1584.990ms
16:35:06.510  	Installing required dependencies...
16:35:10.054  	Uploading build outputs...
16:35:12.872  	Done with "api/spotify.py"
```

Any ideas?

**Comment by novatorem** [comment]
1) Maybe follow this [guide](https://youtu.be/n6d4KHSKqGk?t=615) on set up
2) That's build logs which isn't what we're looking for in this case

Try checking out the functions tab, linked as:
```https://vercel.com/{name}/spotify/{build}/functions``` 

and looks like:

![image](https://user-images.githubusercontent.com/16753077/91338931-b0326680-e7a3-11ea-8178-5499e0e73250.png)

**Comment by Luc1412** [comment]
I got the same issue. Thats my traceback:
```py
[GET] /api/spotify
00:30:09:96
[ERROR]	2020-08-28T22:30:10.226Z	1034fc35-669a-4ff8-bce0-7dc5a6cb52e7	Exception on /api/spotify [GET]
Traceback (most recent call last):
  File "/var/task/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/var/task/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/var/task/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/var/task/flask/_compat.py", line 39, in reraise
    raise value
  File "/var/task/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/var/task/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "./api/spotify.py", line 110, in catch_all
    svg = makeSVG(data)
  File "./api/spotify.py", line 91, in makeSVG
    image = loadImageB64(item["album"]["images"][1]["url"])
IndexError: list index out of range
```

**Comment by novatorem** [comment]
@Luc1412 Does this happen when playing any song or specific ones?

**Comment by RaymondJiang1** [comment]
The guide is really helpful. It helped me set it up and you should also look at it. Thanks to codeSTACKR for helping me set my [website](https://spotify-now-playing-nine.vercel.app/) up! The guide it this => [video](https://youtu.be/n6d4KHSKqGk?t=615). I hope you can deploy it 😄

**Comment by TheOnlyArtz** [comment]
Hey, sorry for the late response, these are the function logs:
```py
16:01:27:05
[ERROR]	2020-08-29T13:01:27.259Z	11d25f05-ac3c-47c4-b3cb-80b638354a5c	Exception on /api/spotify [GET]
Traceback (most recent call last):
  File "/var/task/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/var/task/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/var/task/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/var/task/flask/_compat.py", line 39, in reraise
    raise value
  File "/var/task/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/var/task/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "./api/spotify.py", line 109, in catch_all
    data = nowPlaying()
  File "./api/spotify.py", line 50, in nowPlaying
    token = refreshToken()
  File "./api/spotify.py", line 37, in refreshToken
    return response.json()["access_token"]
KeyError: 'access_token'
```

I've followed the video (which seems to run an outdated version too.)

**Comment by Luc1412** [comment]
> @Luc1412 Does this happen when playing any song or specific ones? 

It happens always.

**Comment by RaymondJiang1** [comment]
> Hey, sorry for the late response, these are the function logs:
> 
> ```python
> 16:01:27:05
> [ERROR]	2020-08-29T13:01:27.259Z	11d25f05-ac3c-47c4-b3cb-80b638354a5c	Exception on /api/spotify [GET]
> Traceback (most recent call last):
>   File "/var/task/flask/app.py", line 2447, in wsgi_app
>     response = self.full_dispatch_request()
>   File "/var/task/flask/app.py", line 1952, in full_dispatch_request
>     rv = self.handle_user_exception(e)
>   File "/var/task/flask/app.py", line 1821, in handle_user_exception
>     reraise(exc_type, exc_value, tb)
>   File "/var/task/flask/_compat.py", line 39, in reraise
>     raise value
>   File "/var/task/flask/app.py", line 1950, in full_dispatch_request
>     rv = self.dispatch_request()
>   File "/var/task/flask/app.py", line 1936, in dispatch_request
>     return self.view_functions[rule.endpoint](**req.view_args)
>   File "./api/spotify.py", line 109, in catch_all
>     data = nowPlaying()
>   File "./api/spotify.py", line 50, in nowPlaying
>     token = refreshToken()
>   File "./api/spotify.py", line 37, in refreshToken
>     return response.json()["access_token"]
> KeyError: 'access_token'
> ```
> 
> I've followed the video (which seems to run an outdated version too.)

it seems like your access_token isn't correct

**Comment by novatorem** [comment]
@Luc1412 your issue is different from @TheOnlyArtz.

@TheOnlyArtz: @RaymondJiang1 is right (thanks for helping!), your access token isn't valid. Try following the steps again and making sure you grab what you need. The video he linked would also be helpful to make sure your values are more or less the same (in terms of format).

@Luc1412 This is odd, because it implies there are no images for the songs you're playing. 
```python
image = loadImageB64(item["album"]["images"][1]["url"])
IndexError: list index out of range
```

Can you change the `[1]` to a `[0]` and see what happens? Although it could be at any point that it breaks apart, so try grabbing all the values from `item`, then `item["album"]`, then `item["album"]["images"]` to see exactly which line breaks.

Basically, replace `image = loadImageB64(item["album"]["images"][1]["url"])` with:

```python
iAlbum = item["album"]
iImages = iAlbum["images"]
iIndex = iImages[0]
iUrl = iIndex["url"]
image = loadImageB64(iUrl)
```

Then run again, and see which line breaks.

**Comment by CatherineOsadciw** [comment]
I'm running into the same issue as @TheOnlyArtz.  Can you clarify what the access_token is referring to?  I've double-checked the client ID and secret ID from Spotify, and have repeated the step for getting the refresh token multiple times.

**Comment by novatorem** [comment]
@CatherineOsadciw Looks like **Luc** fixed their issue, but **TheOnlyArtz** didn't - which is a shame.

<details><summary>The `access_token` comes from the Spotify's authorization code flow:

[Linked](https://developer.spotify.com/documentation/general/guides/authorization-guide/#authorization-code-flow)

</summary>

![image](https://user-images.githubusercontent.com/16753077/92834212-bfbacd80-f3a7-11ea-9840-bd968827405d.png)

</details>

As to the reason, probably a misstep somewhere in the [refresh token section](https://github.com/novatorem/novatorem/blob/master/SetUp.md#refresh-token).

---

Since you already tried multiple times, don't want to leave you at "Try again".

Did the video linked by Raymond at https://github.com/novatorem/novatorem/issues/7#issuecomment-683219785 help?  
Could you share the values you get during the refresh token setup, but obfuscated?

For reference, these are mine (but scrambled):

**{CODE}**
```
AQBJH81WtDsv1oQeXyHpxnBOXWX2Y5fZRIMwoypx6Q1VbW6bDCZqtsdWLC_cvPrta3ZzLXKTpzItOP4H8BZ15WFClRhiljAp6-Tk1JuOdRXqkY-XlIQMdqTezlWmlbXUhWCcRJJm6nhvss8DIlzW5CjdNj1-jwZ0w4OljHwFC2zxESS8Hb2PHkV-tpvfLsAE5MapN-1oMwnBWW9WDBmtoajjAopjnuZC9OUqMpLUr27MWja4nTU
```

**{BASE64}**
```
OGVkZGUzZmE1ZDgwNGI2MmE2MTYwODlhOPNlZTU0MTE6NzkyYTVlMTE5YmM2NDA1OPJlYjYzNDBmZDQ4NIk2MTk=
```

**Curl response**
```
{"access_token":"BQCwQk_0nSOPrn-h1EhSLF0I_dqibBAUaz06Y7wg3nXuUyGAYbiKfurhfikv9KloD0lfjMEZ8tZo5E5qyd4Zu3FpfP5hra7Wxu5CZBdpAUob3ylKJtlCotaLpv8J9uwbCtXYHpwXusYabQnZHQop6uFyCMGc","token_type":"Bearer","expires_in":3600,"refresh_token":"AQACxioi_M0cPQ78xtckIWNKDswGNQQL89umqkmH_jg4IVW-jbhU2e7QdL2gRaCE1c8iiv8udjg5NN5MowcuIa8hS9bRI8WUhOxZDbmpz7YHD3PoesArpZnhRabQ1kER-Ok","scope":"user-read-currently-playing user-read-recently-played"}
```

**Comment by CatherineOsadciw** [comment]
I followed the video as well, but no luck.  Here are my obfuscated values.  At first glance, they all appear to be in the right format.

**{CODE}**
```
6lEh2fnCchC5k0oUJfKtGZKzZgfxiY--O7WrSMv9DXuqMmS9_C8KNZ2NV2rEhNV4hyhyWoCeCiZxqfLATMKHFgKbz_o_aYPYmlAFYQPAZU3qHY7GDTtcKXGwM_WayUuB_BFw4z5w45RQRi0MBb4j2qV0D4U-J4VxGKN_K_IU9MP7ugyQZAKF5dOMeQHTs6GrAyWja1k-21rAbbP7ntoRMsbUdwRfa2ZqYCt-5e-8Ugtil6E_dJM
```

**{BASE64}**
```
MzVJjNGlZM0zI1jdimNZZDgTzMNwjgkm5YzjgTYTOMDNOxZ0Y0kDAmW2dTQm3Y6whTMZiENFIYNwE4jMZND0zTE=
```

**Curl response**
```
{"access_token":"uRLOTdP0AQa9N9Nr4LiPsq0i9dY-hCyxS5s3Dm2c3qBD54hGJ9xOi73fip6jrjVDFWa2U8sD4k63LpIfxZfvaY2CvfT7Utx7UVUUKCX0Ib2eQ1dFaNXOQZz_VCHfTsRttnGJXAvXLL3VwBAssRVJMW8nVFkf43lBmoATXxGXcn","token_type":"Bearer","expires_in":3600,"refresh_token":"IivlPeD_SIwIraSPfNqPL0PomA9NL9od5_IAAhHeLKm8qsZ7oucHq_jiZ6S5Ash_6xRQ49THgbplh9RfeRNoAtpWxpdpQ08zGzHBo2GrZjEeZcQOfbgshRLAVSKSbSc3l0b","scope":"user-read-currently-playing user-read-recently-played"}
```

**Comment by novatorem** [comment]
@CatherineOsadciw Could you implement https://github.com/novatorem/novatorem/commit/44c9ec100b8962c5cf626bc42ae6477dc5b71e14 and share the logs from vercel?

Your values look right, so best bet is looking at what response is sent back.

**Comment by CatherineOsadciw** [comment]
Now it's working!  I implemented your commit, then repeated the same steps for getting the refresh token and updated the environment variable in my deployment accordingly.  No idea what changed if all you added was a try-catch and a log dump!

**Comment by novatorem** [comment]
The try-catch should throw an error if it does break, and I thought it would if it fails to refresh.  
Glad it works now.

Closing, but if anyone else has this issue feel free to reopen!
